core/model.py

```diff
@@ -0,0 +1,230 @@
+                data_load_object[key] = self.__annotations__[key](value)
+            else: # is general type
+                data_load_object[key] = value
+        return data_load_object
+
+    def _get_data_store_object(self) -> dict:
+        data_store_object = {}
+        for key in self.__dict__.keys():
+            if key in _filter_out_store_object_key:
+                continue
+
+            value = self.__dict__[key]
+            if value is None:
+                    continue
+            if self._is_reference_field(key): # is one2many reference
+                element_ids = []
+                for element in value:
+                    if not element._is_saved():
+                        element.save()
+                    element_ids.append(element._id)
+                data_store_object[key] = element_ids
+            elif isinstance(value, Model): # is many2one reference
+                if not value._is_saved():
+                    value.save()
+                data_store_object[key] = value._id
+            elif isinstance(value, Enum): # is enum
+                data_store_object[key] = value.value
+            else: # is general type
+                data_store_object[key] = value
+        return data_store_object
+
+    def _is_saved(self) -> bool:
+        return hasattr(self, '_id') and self._id
+    
+    def _is_reference_field(self, field: str) -> bool:
+        return field in self.__annotations__ and len(get_args(self.__annotations__[field])) == 1 and isinstance(get_args(self.__annotations__[field])[0](), Model)
+    
+    def _get_reference_model_of_field(self, field: str):
+        return get_args(self.__annotations__[field])[0]
+
+    @classmethod
+    def _get_fields(cls):
+        return list(cls.__annotations__.keys()) + _prefilled_fields
+
+    def save(self: T) -> T:
+        current_utc_time = datetime.utcnow()
+        if self._is_saved():
+            self.updated_at = current_utc_time
+            self._database.update(self._get_collection_name(), self._id, self._get_data_store_object())
+        else:
+            self.created_at = current_utc_time
+            self.updated_at = current_utc_time
+            saved_data_id = self._database.create(self._get_collection_name(), self._get_data_store_object())
+            self._id = saved_data_id
+        return self
+
+    def update(self: T, data: dict) -> T:
+        for key in data:
+            if key in self.__annotations__:
+                setattr(self, key, data[key])
+        self.save()
+        return self
+
+    @classmethod
+    def _get_collection_name(cls: T):
+        object_name = cls.__name__
+        snake_case_name = re.sub(r'(?<!^)(?=[A-Z])', '_', object_name).lower()
+        return snake_case_name
+    
+    @classmethod
+    def get(cls: T, id: Union[str, ObjectId]) -> T:
+        class_instance = cls()
+        return cls(class_instance._get_data_load_object(class_instance._database.get(cls._get_collection_name(), id)))
```

Comment:
This looks awfully over-complicated. You create a fake object `class_instance` just to use it to fetch a real object?
Shouldn't this be something like `return cls(cls._database.get(cls.__get_collection_name(), id))`.
I guess it would be convenient if you could get a connection to the database without creating fake objects, i.e. if you could create a database object and give it a collection name, e.g.
`return cls(Database(cls._get_collection_name()).get(id))`?

Reply:
Hmm. I still need `_get_data_load_object` to map the data from database to actual object. I don't know if I change `_get_data_load_object ` to be classmethod, if it makes more cleaner since it does not require fake object to be created?.

Reply:
It makes sense to have `_get_data_load_object` to be classmethod. `_database` can probably be global as it does not represent some exact collection, but is the whole database.
Maybe it would be better to have `_database` as class attribute where it already points to the collection of that class, i.e. `cls._database` would already be a required collection as campaigns or reports?